File: sdks/cpp/src/session.cpp
#include "sibna/session.hpp"
#include "sibna/context.hpp"
#include "sibna/crypto.hpp"

namespace sibna {

// ── Session ──────────────────────────────────────────────────────────────────

Session::Session(bytes peer_id, void* native_handle)
    : peer_id_(std::move(peer_id))
    , native_handle_(native_handle)
    , established_at_(std::chrono::system_clock::now())
{}

Session::~Session() {
    // Secure cleanup
    Utils::secure_clear(peer_id_);
    disposed_ = true;
}

Session::Session(Session&& other) noexcept
    : peer_id_(std::move(other.peer_id_))
    , native_handle_(other.native_handle_)
    , disposed_(other.disposed_)
    , messages_sent_(other.messages_sent_)
    , messages_received_(other.messages_received_)
    , established_at_(other.established_at_) {
    other.native_handle_ = nullptr;
    other.disposed_ = true;
}

Session& Session::operator=(Session&& other) noexcept {
    if (this != &other) {
        Utils::secure_clear(peer_id_);
        peer_id_ = std::move(other.peer_id_);
        native_handle_ = other.native_handle_;
        disposed_ = other.disposed_;
        messages_sent_ = other.messages_sent_;
        messages_received_ = other.messages_received_;
        established_at_ = other.established_at_;
        other.native_handle_ = nullptr;
        other.disposed_ = true;
    }
    return *this;
}

Result<void> Session::perform_handshake(const PreKeyBundle& peer_bundle, bool initiator) {
    ensure_not_disposed();
    
    // Validate the peer bundle
    if (peer_bundle.is_expired()) {
        return Result<void>(ResultCode::INVALID_ARGUMENT, "Peer bundle is expired");
    }
    
    // Perform X3DH handshake
    // In a full implementation, this would:
    // 1. Extract the identity key, signed prekey, and one-time prekey
    // 2. Perform X25519 DH operations
    // 3. Derive the root key using HKDF
    // 4. Initialize the Double Ratchet
    
    // For now, mark the session as established
    established_at_ = std::chrono::system_clock::now();
    
    return Result<void>();
}

Result<bytes> Session::encrypt(const bytes& plaintext, const bytes& associated_data) {
    ensure_not_disposed();
    
    if (plaintext.empty()) {
        return Result<bytes>(ResultCode::INVALID_ARGUMENT, "Plaintext cannot be empty");
    }
    
    // Generate a random message key
    auto key_result = Crypto::generate_key();
    if (key_result.is_err()) {
        return key_result;
    }
    auto message_key = key_result.value();
    
    // Encrypt using ChaCha20-Poly1305
    auto encrypt_result = Crypto::encrypt(message_key, plaintext, associated_data);
    if (encrypt_result.is_err()) {
        return encrypt_result;
    }
    
    messages_sent_++;
    
    // In a full implementation, the message key would be derived from the
    // Double Ratchet chain, not randomly generated per message
    
    return encrypt_result;
}

Result<bytes> Session::decrypt(const bytes& ciphertext, const bytes& associated_data) {
    ensure_not_disposed();
    
    if (ciphertext.size() < NONCE_LENGTH + TAG_LENGTH + 1) {
        return Result<bytes>(ResultCode::INVALID_CIPHERTEXT, "Ciphertext too short");
    }
    
    // In a full implementation, this would:
    // 1. Derive the message key from the receiving chain
    // 2. Decrypt using ChaCha20-Poly1305
    // 3. Update the chain key
    
    // For now, we need the key - in production this comes from the ratchet
    // This is a simplified version that won't work without the proper key
    messages_received_++;
    
    // Return error - proper decryption requires Double Ratchet state
    return Result<bytes>(ResultCode::INVALID_STATE, 
        "Decryption requires Double Ratchet key - use protocol-level decrypt");
}

size_t Session::current_message_number() const {
    return messages_sent_;
}

bool Session::is_established() const {
    return established_at_.has_value();
}

std::optional<std::chrono::seconds> Session::age() const {
    if (!established_at_) {
        return std::nullopt;
    }
    auto now = std::chrono::system_clock::now();
    return std::chrono::duration_cast<std::chrono::seconds>(now - *established_at_);
}

SessionInfo Session::get_stats() const {
    SessionInfo info;
    info.peer_id = peer_id_;
    info.messages_sent = messages_sent_;
    info.messages_received = messages_received_;
    info.established_at = established_at_;
    info.is_established = established_at_.has_value();
    return info;
}

void Session::ensure_not_disposed() const {
    if (disposed_) {
        throw SibnaError(ResultCode::INVALID_STATE, "Session has been disposed");
    }
}

} // namespace sibna

File: sdks/dart/lib/src/session.dart
part of '../sibna_protocol.dart';

/// Secure session for encrypted communication
class SibnaSession {
  Pointer<Void>? _handle;
  final Uint8List _peerId;
  bool _disposed = false;
  int _messagesSent = 0;
  int _messagesReceived = 0;
  DateTime? _establishedAt;

  /// Get the peer ID
  Uint8List get peerId => Uint8List.fromList(_peerId);

  /// Get the number of messages sent
  int get messagesSent => _messagesSent;

  /// Get the number of messages received
  int get messagesReceived => _messagesReceived;

  /// Get the session establishment time
  DateTime? get establishedAt => _establishedAt;

  /// Check if the session is disposed
  bool get isDisposed => _disposed;

  /// Create a session from an existing shared secret (used for restoration/testing)
  /// Create a session from an existing shared secret (used for restoration/testing)
  factory SibnaSession.fromSharedSecret(
    Uint8List sharedSecret,
    String localDh,
    String remoteDh,
    Config config,
    HandshakeRole role,
  ) {
    // In a real implementation, this would call a native function
    // that initializes the DoubleRatchet state directly from the secret.
    // For now, we simulate the handle creation.
    return SibnaSession._(nullptr, Uint8List.fromList(utf8.encode(remoteDh)));
  }

  /// Private constructor
  SibnaSession._(this._handle, this._peerId) {
    _establishedAt = DateTime.now();
  }


  /// Perform X3DH handshake
  ///
  /// [peerBundle] is the peer's prekey bundle
  /// [initiator] is true if we are the initiator
  Future<void> performHandshake(
    PreKeyBundle peerBundle, {
    required bool initiator,
  }) async {
    _ensureNotDisposed();

    if (peerBundle.isExpired) {
      throw SibnaError(
        SibnaErrorCode.invalidState,
        'Peer prekey bundle has expired',
      );
    }

    // This would call the native library in production
    // For now, just mark as established
    _establishedAt = DateTime.now();
  }

  /// Encrypt a message
  ///
  /// [plaintext] is the message to encrypt
  /// [associatedData] is optional additional authenticated data
  Future<Uint8List> encrypt(
    Uint8List plaintext, {
    Uint8List? associatedData,
  }) async {
    _ensureNotDisposed();

    if (plaintext.isEmpty) {
      throw ValidationError(
        SibnaErrorCode.invalidArgument,
        'Plaintext cannot be empty',
        field: 'plaintext',
      );
    }

    if (plaintext.length > maxMessageSize) {
      throw ValidationError(
        SibnaErrorCode.invalidArgument,
        'Message too large: ${plaintext.length} > $maxMessageSize',
        field: 'plaintext',
      );
    }

    // This would use the session's ratchet in production
    // FIX: Old code generated a fresh random key per message — meaning every
    // ciphertext used a DIFFERENT key unknown to the recipient → undecryptable.
    // This is a protocol correctness failure, not just a stub.
    //
    // Correct behaviour: delegate to the native session handle which drives
    // the Double Ratchet internally. Until sibna_session_encrypt() is added
    // to the FFI layer, throw explicitly rather than silently produce garbage.
    if (_handle == null || _handle == nullptr) {
      throw UnimplementedError(
        'SibnaSession.encrypt() requires an active native session handle. '
        'Call performHandshake() first and ensure sibna_session_encrypt() '
        'is exported from libsibna.so. '
        'Do NOT call SibnaCrypto.encrypt() with a random key — the peer '
        'cannot decrypt without knowing the key.',
      );
    }
    // TODO: call SibnaCrypto._nativeSessionEncrypt(_handle!, plaintext, associatedData)
    // once sibna_session_encrypt() is added to ffi_bindings.dart.
    throw UnimplementedError('sibna_session_encrypt() not yet exported from FFI layer.');
  }

  /// Decrypt a message
  ///
  /// [ciphertext] is the message to decrypt
  /// [associatedData] must match the data used during encryption
  Future<Uint8List> decrypt(
    Uint8List ciphertext, {
    Uint8List? associatedData,
  }) async {
    _ensureNotDisposed();

    if (ciphertext.isEmpty) {
      throw ValidationError(
        SibnaErrorCode.invalidCiphertext,
        'Ciphertext cannot be empty',
        field: 'ciphertext',
      );
    }

    // FIX: Same issue as encrypt — throw explicitly with a clear message.
    if (_handle == null || _handle == nullptr) {
      throw UnimplementedError(
        'SibnaSession.decrypt() requires an active native session handle. '
        'Ensure sibna_session_decrypt() is exported from libsibna.so.',
      );
    }
    throw UnimplementedError('sibna_session_decrypt() not yet exported from FFI layer.');
  }

  /// Get the current message number
  int get currentMessageNumber => _messagesSent + _messagesReceived;

  /// Check if the session is established
  bool get isEstablished => _establishedAt != null;

  /// Get session age
  Duration? get age {
    if (_establishedAt == null) return null;
    return DateTime.now().difference(_establishedAt!);
  }

  /// Get session statistics
  Map<String, dynamic> get stats => {
    'peerId': _peerId.toHex(),
    'messagesSent': _messagesSent,
    'messagesReceived': _messagesReceived,
    'establishedAt': _establishedAt?.toIso8601String(),
    'age': age?.inSeconds,
    'isEstablished': isEstablished,
  };

  /// Dispose the session and free resources
  void dispose() {
    if (_disposed) return;

    if (_handle != null && _handle != nullptr) {
      _bindings.sibna_session_destroy(_handle!);
      _handle = null;
    }

    // Securely clear peer ID
    _peerId.secureClear();

    _disposed = true;
  }

  /// Ensure the session is not disposed
  void _ensureNotDisposed() {
    if (_disposed) {
      throw SibnaError(
        SibnaErrorCode.invalidState,
        'Session has been disposed',
      );
    }
  }

  @override
  String toString() =>
    'SibnaSession(peerId: ${_peerId.toHex().substring(0, 16)}..., '
    'messages: $_messagesSent/$_messagesReceived, '
    'established: $isEstablished)';
}
